include/uart.h

```diff
@@ -29,6 +29,157 @@ extern "C" {
+ *   |      |       |              |  |     |     UART_RX_RDY
+ *   |      |       |              |  |     |     UART_RX_BUF_REQUEST
+ *   |      |       |              |  |     |
+ *   |      |       |              |  |     UART_RX_RDY (Timeout)
+ *   |      |       |              |  |
+ *   |      |       |              |  UART_RX_RDY (Timeout)
+ *   |      |       |              |
+ *   |      |       |              UART_RX_RDY (Buffer 1 full)
+ *   |      |       |              UART_RX_BUF_REQUEST
+ *   |      |       |
+ *   |      |       UART_RX_RDY (Buffer 0 full)
+ *   |      |       UART_RX_BUF_REQUEST (Started transmission to second buffer
+ *   |      |                            need another buffer)
+ *   |      |
+ *   |      UART_RX_RDY (Timeout occurred, received n bytes)
+ *   |
+ *   UART_RX_BUF_REQUEST (Started transmission to first buffer,
+ *                        need second buffer)
+ *
+ */
+enum uart_event_type {
+	/** @brief Whole TX buffer was transmitted. */
+	UART_TX_DONE,
+	/** @brief TX timed out.
+	 *
+	 * When flow control is enabled, there is a possibility that TX transfer
+	 * won't finish.
+	 */
+	UART_TX_TIMEOUT,
+	/**
+	 * @brief Received data is ready for processing.
+	 *
+	 * This event is generated in two cases:
+	 * - When RX timeout occurred, and data was stored in provided buffer.
+	 * - When provided buffer is full.
+	 */
+	UART_RX_RDY,
+	/**
+	 * @brief Driver requests next buffer for continuous transmission.
+	 *
+	 * This event is triggered when receiving has started for a new buffer.
+	 * For continuous reliable receiving, user should provide another RX
+	 * buffer in response to this event, using uart_rx_buf_rsp() function.
+	 *
+	 * If uart_rx_buf_rsp() function is not called, receiving will stop
+	 * after filling current buffer and UART_RX_DONE will be generated.
+	 */
+	UART_RX_BUF_REQUEST,
+	/**
+	 * @brief Receiving has stopped and provided buffer is full.
+	 *
+	 * If no second buffer was provided, receiving will stop after filling
+	 * current buffer. After that event, receiving can be enabled again
+	 * using rx_en() function.
+	 */
+	UART_RX_DONE,
+	/**
+	 * @brief Buffer is no longer used by UART driver.
+	 *
+	 * This event occurs after rx_disable() function is called, for every
+	 * buffer passed to UART driver, it indicates that buffer can be freed
+	 * and is no longer used.
+	 */
+	UART_RX_BUF_RELEASED,
+	/**
+	 * @brief RX has been disabled and can be reenabled.
+	 *
+	 * After rx_disable() function is called, driver may need some time to
+	 * stop the transmission, after releasing buffers this event is
+	 * generated and receiving can be enabled again using rx_en() function.
+	 */
+	UART_RX_DISABLED,
+	/**
+	 * @brief Break interrupt
+	 *
+	 * A break interrupt was received. This happens when the serial input
+	 * is held at a logic '0' state for longer than the sum of
+	 * start time + data bits + parity + stop bits.
+	 */
+	UART_BREAK,
```

Comment:
what about making `UART_ERROR`  and type of error as event parameter?

Reply:
I'm not sure about this comment. @nordic-krch, do you suggest to report UART BREAK condition as an error? I'm not sure that's a good idea. BREAK is a kind of out-of-band data, but not an error per se.